libraries/AP_UAVCAN/AP_UAVCAN.cpp

```diff
@@ -419,172 +119,182 @@ AP_UAVCAN::~AP_UAVCAN()
-    if (gnss_fix_start_res < 0) {
-        debug_uavcan(1, "UAVCAN GNSS subscriber start problem\n\r");
-        return false;
-    }
-
-    uavcan::Subscriber<uavcan::equipment::gnss::Auxiliary> *gnss_aux;
-    gnss_aux = new uavcan::Subscriber<uavcan::equipment::gnss::Auxiliary>(*node);
-    const int gnss_aux_start_res = gnss_aux->start(gnss_aux_cb_arr[_uavcan_i]);
-    if (gnss_aux_start_res < 0) {
-        debug_uavcan(1, "UAVCAN GNSS Aux subscriber start problem\n\r");
-        return false;
-    }
-
-    uavcan::Subscriber<uavcan::equipment::ahrs::MagneticFieldStrength> *magnetic;
-    magnetic = new uavcan::Subscriber<uavcan::equipment::ahrs::MagneticFieldStrength>(*node);
-    const int magnetic_start_res = magnetic->start(magnetic_cb_arr[_uavcan_i]);
-    if (magnetic_start_res < 0) {
-        debug_uavcan(1, "UAVCAN Compass subscriber start problem\n\r");
-        return false;
-    }
-
-    uavcan::Subscriber<uavcan::equipment::ahrs::MagneticFieldStrength2> *magnetic2;
-    magnetic2 = new uavcan::Subscriber<uavcan::equipment::ahrs::MagneticFieldStrength2>(*node);
-    const int magnetic_start_res_2 = magnetic2->start(magnetic_cb_2_arr[_uavcan_i]);
-    if (magnetic_start_res_2 < 0) {
-        debug_uavcan(1, "UAVCAN Compass for multiple mags subscriber start problem\n\r");
-        return false;
+    int start_res = _node->start();
+    if (start_res < 0) {
+        debug_uavcan(1, "UAVCAN: node start problem, error %d\n\r", start_res);
+        return;
     }
 
-    uavcan::Subscriber<uavcan::equipment::air_data::StaticPressure> *air_data_sp;
-    air_data_sp = new uavcan::Subscriber<uavcan::equipment::air_data::StaticPressure>(*node);
-    const int air_data_sp_start_res = air_data_sp->start(air_data_sp_cb_arr[_uavcan_i]);
-    if (air_data_sp_start_res < 0) {
-        debug_uavcan(1, "UAVCAN Baro subscriber start problem\n\r");
-        return false;
-    }
-    
-    uavcan::Subscriber<uavcan::equipment::air_data::StaticTemperature> *air_data_st;
-    air_data_st = new uavcan::Subscriber<uavcan::equipment::air_data::StaticTemperature>(*node);
-    const int air_data_st_start_res = air_data_st->start(air_data_st_cb_arr[_uavcan_i]);
-    if (air_data_st_start_res < 0) {
-        debug_uavcan(1, "UAVCAN Temperature subscriber start problem\n\r");
-        return false;
-    }
-    
-    uavcan::Subscriber<uavcan::equipment::power::BatteryInfo> *battery_info_st;
-    battery_info_st = new uavcan::Subscriber<uavcan::equipment::power::BatteryInfo>(*node);
-    const int battery_info_start_res = battery_info_st->start(battery_info_st_cb_arr[_uavcan_i]);
-    if (battery_info_start_res < 0) {
-        debug_uavcan(1, "UAVCAN BatteryInfo subscriber start problem\n\r");
-        return false;
-    }
+    //Roundup all subscribers from supported drivers
+    AP_GPS::uavcan_init_callback(this);
+    Compass::uavcan_init_callback(this);
+    AP_Baro::uavcan_init_callback(this);
+    AP_BattMonitor::uavcan_init_callback(this);
```

Comment:
This was the only way I was able to figure out reasonably clean way to initialise before init methods of each sensors are called.

Reply:
I think you didn't get my comment. What I was saying was that this should be `AP_GPS_UAVCAN::init_callback(this);`, etc.; be in the actual UAVCAN driver instead of the frontend.